pywbem_mock/_mainprovider.py

```diff
@@ -0,0 +1,4208 @@
+except ImportError:
+    # time.clock() was deprecated in Python 3.3
+    from time import clock as delta_time
+# pylint: enable=ungrouped-imports
+
+
+from pywbem import CIMClass, CIMClassName, \
+    CIMInstance, CIMInstanceName, CIMQualifierDeclaration, CIMError, \
+    CIM_ERR_NOT_FOUND, CIM_ERR_INVALID_PARAMETER, CIM_ERR_INVALID_CLASS, \
+    CIM_ERR_ALREADY_EXISTS, CIM_ERR_INVALID_NAMESPACE, \
+    CIM_ERR_INVALID_ENUMERATION_CONTEXT, CIM_ERR_NOT_SUPPORTED, \
+    CIM_ERR_QUERY_LANGUAGE_NOT_SUPPORTED, CIM_ERR_NAMESPACE_NOT_EMPTY, \
+    CIM_ERR_FAILED
+from pywbem._nocasedict import NocaseDict
+from pywbem._utils import _format
+
+from ._resolvermixin import ResolverMixin
+
+# Default Max_Object_Count for Fake Server if not specified by request
+_DEFAULT_MAX_OBJECT_COUNT = 100
+
+# Maximum Open... timeout if not set by request
+OPEN_MAX_TIMEOUT = 40
+
+# Per DSP0200, the default behavior for EnumerateInstance DeepInheritance
+# if not set by server.  Default is True.
+DEFAULT_DEEP_INHERITANCE = True
+
+
+# None of the request method names conform since they are camel case
+# pylint: disable=invalid-name
+
+
+class MainProvider(ResolverMixin, object):
+    """
+    A CIM provider in the mock support of pywbem creates WBEM server responses
+    to operations defined in DSP0200 and implements them using a CIM
+    repository. Because the CIM repository has a uniform interface (defined by
+    BaseRepository), most operations can be implemented in a generic way, and
+    only those operations that create, delete or modify CIM instances have a
+    need to have user-defined providers.
+
+    The MainProvider class is the main CIM provider in the mock support of
+    pywbem and handles all those operations that do not have a need to be
+    implementable by users. Specifically, it implements:
+
+    * All operations on CIM classes.
+    * All operations on CIM qualifier declarations.
+    * All read operations on CIM instances (i.e. get, traditional and
+      pull enumerations, and association operations).
+
+    Operations that create, delete or modify CIM instances or that invoke
+    static or non-static methods have a default implementation in the
+    BaseInstanceProvider class and can be overwritten by users on a per class
+    and per operation basis.
+
+    The MainProvider class also includes methods to add namespaces, remove
+    namespaces and get a list of namespaces in the CIM repository.
+
+    For more details, see mocksupport.rst.
+    """
+
+    def __init__(self, conn, cimrepository):
+        # pylint: disable=line-too-long
+        """
+        Parameters:
+            conn(:class:`~pywbem_mock.FakedWBEMConnection`):
+              The current instance of the connection from which some
+              variables are extracted for the repository including
+              the host and disable_pull_operations attributes.
+
+            repository(:class:`~pywbem_mock.BaseRepository` or derived from `~pywbem_mock.BaseRepository`):
```

Comment:
Since `BaseRepository` is an abstract base class, it is always going to be a derived class. Change the type to just `(derived from `~pywbem_mock.BaseRepository`)`.

Reply:
DONE